phoenix/app/api_products/lib/api_products/plugs/get_product_plug.ex

```diff
@@ -0,0 +1,24 @@
+defmodule ApiProducts.GetProductPlug do
+  use ApiProductsWeb, :controller
+  import Plug.Conn
+
+  alias ApiProducts.Management
+
+  def init(props) do
+    props
+  end
+
+  def call(conn, _) do
+    product = Management.get_product(conn.params["id"])
+
+    case product do
+      nil ->
```

Comment:
Um ponto importante sobre pattern match em geral no elixir é que a ordem é de extrema importância. Isso pois na execução do código sempre a primeira clausula que o pattern match identificar será a executada.
Digo isso pois esse case apresenta 2 problemas:

1. E se a variável product não for nem nil nem uma struct de Products? Isso pode gerar uma execução indesejada, por isso é de suma importancia criar um case genérico que resgate todas as possibilidades não previstas, pois assim o código fica mais resistente a mudanças futuras.
2. A ordem das clausulas de verificação está estranha, o que inicialmente não é problema já que ela é rígida, mas se adicionar-mos uma clausula genérica para pegar os casos desconhecidos e não nos preocupar-mos com a ordem, corre-se o risco da clausula genérica capturar o fluxo inteiro e criar comportamentos indesejados.

Enfim, seria melhor que, acredito eu, esse nil fosse uma clausula genérica e a ordem das clausulas fossem invertidas